lib/programmemory.cpp

```diff
@@ -147,57 +149,87 @@ void ProgramMemory::setContainerSizeValue(const Token* expr, MathLib::bigint val
 }
 
 void ProgramMemory::setUnknown(const Token* expr) {
-    mValues[expr].valueType = ValueFlow::Value::ValueType::UNINIT;
+    copyOnWrite();
+
+    (*mValues)[expr].valueType = ValueFlow::Value::ValueType::UNINIT;
 }
 
 bool ProgramMemory::hasValue(nonneg int exprid)
 {
-    return mValues.find(exprid) != mValues.end();
+    return mValues->find(exprid) != mValues->end();
 }
 
 const ValueFlow::Value& ProgramMemory::at(nonneg int exprid) const {
-    return mValues.at(exprid);
+    return mValues->at(exprid);
 }
 ValueFlow::Value& ProgramMemory::at(nonneg int exprid) {
-    return mValues.at(exprid);
+    copyOnWrite();
+
+    return mValues->at(exprid);
 }
 
 void ProgramMemory::erase_if(const std::function<bool(const ExprIdToken&)>& pred)
 {
-    for (auto it = mValues.begin(); it != mValues.end();) {
+    if (mValues->begin() == mValues->end())
+        return;
+
+    // TODO: how to delay until we actuallly modify?
+    copyOnWrite();
+
+    for (auto it = mValues->begin(); it != mValues->end();) {
         if (pred(it->first))
-            it = mValues.erase(it);
+        {
+            it = mValues->erase(it);
+        }
         else
             ++it;
     }
 }
 
 void ProgramMemory::swap(ProgramMemory &pm)
 {
-    mValues.swap(pm.mValues);
+    copyOnWrite();
+    pm.copyOnWrite();
+
+    mValues->swap(*pm.mValues);
 }
 
 void ProgramMemory::clear()
 {
-    mValues.clear();
+    copyOnWrite();
+
+    mValues->clear();
 }
 
 bool ProgramMemory::empty() const
 {
-    return mValues.empty();
+    return mValues->empty();
 }
 
 void ProgramMemory::replace(ProgramMemory pm)
 {
-    for (auto&& p : pm.mValues) {
-        mValues[p.first] = std::move(p.second);
+    copyOnWrite();
+
+    for (auto&& p : (*pm.mValues)) {
+        (*mValues)[p.first] = std::move(p.second);
     }
 }
 
 void ProgramMemory::insert(const ProgramMemory &pm)
 {
+    copyOnWrite();
+
     for (auto&& p : pm)
-        mValues.insert(p);
+        mValues->insert(p);
+}
+
+void ProgramMemory::copyOnWrite()
+{
+    if (!mCopyValues)
```

Comment:
There is no need for the `mCopyValues` variable. Just check `mValues.use_count() == 1` instead.

Reply:
Will do.

Reply:
Done.